### PR TITLE
Adds CSR Builder (Redux of #1927)

### PR DESCRIPTION
**Pull request by sigmavirus24**
There's a lot of history in #1927 so go read that for review comments and such before looking over here :wink:

### CODE REVIEW

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 1]
```diff
@@ -686,6 +776,68 @@ def cmac_algorithm_supported(self, algorithm):
     def create_cmac_ctx(self, algorithm):
         return _CMACContext(self, algorithm)
 
+    def create_x509_csr(self, builder, private_key, algorithm):
+        # TODO: check type of private key parameter.
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Not every private_key type is guaranteed to have an `_evp_pkey` right now (and, in fact, only RSA does). There has been past discussion in #1231 about having all our OpenSSL asym objects take evp_pkey, so maybe we should do that so this doesn't need to know anything about type. Alternately this will need to check type and create an evp_pkey for DSA/EC.

Either way, we'll need at least one CSR generation for EC and DSA to confirm this works (which right now it won't).

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -56,6 +56,96 @@
...
+def _encode_name(backend, attributes):
+    subject = backend._lib.X509_NAME_new()
+    subject = backend._ffi.gc(subject, backend._lib.X509_NAME_free)
+    for attribute in attributes:
+        value = attribute.value.encode('ascii')
+        obj = _txt2obj(backend, attribute.oid.dotted_string)
+        res = backend._lib.X509_NAME_add_entry_by_OBJ(
+            subject,
+            obj,
+            backend._lib.MBSTRING_ASC,
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Pulling the discussion from #1927 over to here, we need to do some experimentation to determine whether this should be MBSTRING_UTF8 (and whether value should therefore be encoded utf8 rather than ascii).

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (sigmavirus24) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Should we add xfail tests for EC and DSA so when that's implemented, we'll have tests that will pass without doing extra work?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: I wouldn't even xfail them. Let them fail CI since that will remind us that this PR isn't ready yet (and we can rebase/restart builds after the evp_pkey improvement merges)

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (AndreLouisCaron) [inline src/cryptography/hazmat/backends/openssl/backend.py]: @reaperhulk Would that have an impact on compatibility with older software?  Should we instead try to encode it to ASCII and then fall back to UTF-8 if there are non-ASCII characters?

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (AndreLouisCaron) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Are you saying this PR should wait until another refactoring occurs or should we add the tests to make the EC and DSA cases work in this PR first?

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: We should add tests for EC and DSA. Those tests will fail until the evp_pkey change is merged though.

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Is there specific software known to fail with UTF8?

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: I'm also open to having some branches here, based on private key type. Which do you feel is better @reaperhulk ?

File: src/cryptography/x509.py [hunk 3]
```diff
@@ -1375,3 +1375,43 @@ def extensions(self):
...
+
+    def add_extension(self, extension, critical=False):
+        """
+        Adds an X.509 extension to the certificate request.
+        """
+        if isinstance(extension, BasicConstraints):
+            extension = Extension(OID_BASIC_CONSTRAINTS, critical, extension)
+        else:
+            raise ValueError('Unsupported X.509 extension.')
+        for e in self._extensions:
```
Comment (alex) [inline src/cryptography/x509.py]: Can you add a `TODO: this is quadratic in the number of extensions` comment. We don't need to fix now, I just want it to be obvious in the ftureu

File: src/cryptography/x509.py [hunk 4]
```diff
@@ -1375,3 +1375,43 @@ def extensions(self):
...
+        return CertificateSigningRequestBuilder(name, self._extensions)
+
+    def add_extension(self, extension, critical=False):
+        """
+        Adds an X.509 extension to the certificate request.
+        """
+        if isinstance(extension, BasicConstraints):
+            extension = Extension(OID_BASIC_CONSTRAINTS, critical, extension)
+        else:
+            raise ValueError('Unsupported X.509 extension.')
```
Comment (alex) [inline src/cryptography/x509.py]: `NotImplementedError` here please.

File: src/cryptography/x509.py [hunk 5]
```diff
@@ -1375,3 +1375,43 @@ def extensions(self):
...
+        """
+        Sets the certificate requestor's distinguished name.
+        """
+        if not isinstance(name, Name):
+            raise TypeError('Expecting x509.Name object.')
+        if self._subject_name is not None:
+            raise ValueError('The subject name may only be set once.')
+        return CertificateSigningRequestBuilder(name, self._extensions)
+
+    def add_extension(self, extension, critical=False):
```
Comment (alex) [inline src/cryptography/x509.py]: @reaperhulk how do you feel about having a default on critical?

Comment (reaperhulk) [inline src/cryptography/x509.py]: I'd prefer we drop the default and require that users explicitly pass the boolean. Our API design has consistently favored explicit over implicit defaults and I don't see a compelling reason to change that for adding an extension (even though False will be far and away the most common value).

**Comment by alex** [comment]
the `_evp_pkey` stuff is done now

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Now that evp_pkey change is merged we can just add the tests and remove the todo. No branches needed

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: I wouldn't expect compatibility problems, no. PyOpenSSL (among many others) use this already. I also confirmed UTF8 (in my test case an emoji) is properly handled when doing:

```python
    for attribute in attributes:
        value = attribute.value.encode('utf8')
        obj = _txt2obj(backend, attribute.oid.dotted_string)
        res = backend._lib.X509_NAME_add_entry_by_OBJ(
            subject,
            obj,
            backend._lib.MBSTRING_UTF8,
            value,
            -1, -1, 0,
        )
        assert res == 1
```

We should add a test to confirm we can parse unicode back out properly, but it looks like this is the right way forward.

File: tests/test_x509.py [hunk 6]
```diff
@@ -646,6 +647,117 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
...
+        builder = x509.CertificateSigningRequestBuilder().subject_name(
+            x509.Name([
+                x509.NameAttribute(x509.OID_COUNTRY_NAME, u'US'),
+                x509.NameAttribute(x509.OID_STATE_OR_PROVINCE_NAME, u'Texas'),
+                x509.NameAttribute(x509.OID_LOCALITY_NAME, u'Austin'),
+                x509.NameAttribute(x509.OID_ORGANIZATION_NAME, u'PyCA'),
+                x509.NameAttribute(x509.OID_COMMON_NAME, u'cryptography.io'),
+            ])
+        )
+        with pytest.raises(ValueError):
```
Comment (alex) [inline tests/test_x509.py]: This should be `NotImplementedError`

File: tests/test_x509.py [hunk 7]
```diff
@@ -646,6 +647,117 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
...
+    def test_add_unsupported_extension(self, backend):
+        builder = x509.CertificateSigningRequestBuilder().subject_name(
+            x509.Name([
+                x509.NameAttribute(x509.OID_COUNTRY_NAME, u'US'),
+                x509.NameAttribute(x509.OID_STATE_OR_PROVINCE_NAME, u'Texas'),
+                x509.NameAttribute(x509.OID_LOCALITY_NAME, u'Austin'),
+                x509.NameAttribute(x509.OID_ORGANIZATION_NAME, u'PyCA'),
+                x509.NameAttribute(x509.OID_COMMON_NAME, u'cryptography.io'),
+            ])
+        )
```
Comment (alex) [inline tests/test_x509.py]: You can drop the subject name from this example

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (sigmavirus24) [inline src/cryptography/hazmat/backends/openssl/backend.py]: So we'd add these tests to `tests/hazmat/backends/test_openssl.py`, correct? We'll need one for each of `RSA`, `EC`, and `DSA`, right?

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (sigmavirus24) [inline src/cryptography/hazmat/backends/openssl/backend.py]: > We should add a test to confirm we can parse unicode back out properly, but it looks like this is the right way forward.

Could you provide some guidance? There aren't presently any tests in `tests/hazmat/backends/test_openssl.py` for `create_x509_csr`, and I don't see anything that parses a request as part of this, but I'm guessing that was added before this PR?

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 1 above)
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/backend.py]: No, tests to make sure all types of keys work should just be in `test_x509` -- these are tests of the common interface, not any one implementation

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: Yep, just add decorators for requires_backend_interface so the tests only run if the given backend supports EC or DSA. Then load a DSA/EC key in the test and sign with it.

File: src/cryptography/hazmat/backends/openssl/backend.py (hunk 2 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/backend.py]: This test can live in test_x509 as well since it should be backend agnostic (we want to make sure any backend we implement properly handles unicode after all). You'll just create a builder and sign a request, then parse the results using `x509.load_pem_x509_csr` and confirm that the subject (which should contain some unicode characters) round tripped successfully.

**Comment by alex** [comment]
So at this point the only two outstanding items are:

- Tests for DSA and EC private keys
- Encoding Names as UTF-8

@reaperhulk do you concur^ ?

Thanks so much to @sigmavirus24 and @AndreLouisCaron for all your work on this!

**Comment by reaperhulk** [comment]
Yep, I'd like to give it one more review once those are addressed but it looks like all comments thus far are covered.

**Comment by sigmavirus24** [comment]
With any luck the tests should now pass.

File: tests/hazmat/backends/test_openssl.py [hunk 8]
```diff
@@ -453,6 +454,28 @@ class FakeAlgorithm(object):
...
+
+        with pytest.raises(NotImplementedError):
+            backend.create_x509_csr(object(), private_key, hashes.SHA1())
+
+    @pytest.mark.skipif(
+        backend._lib.OPENSSL_VERSION_NUMBER >= 0x10001000,
+        reason="Requires an older OpenSSL. Must be < 1.0.1"
+    )
+    def test_unsupported_ec_keys(self):
+        private_key = ec.generate_private_key(ec.SECT283K1(), backend)
```
Comment (reaperhulk) [inline tests/hazmat/backends/test_openssl.py]: This also needs a `_skip_curve_unsupported`

**Comment by reaperhulk** [comment]
I'd like to see a test that encodes a multi-byte unicode char (e.g. u"\U0001F37A") into a CSR, then parses the CSR (via `x509.load_pem_x509_csr`) and confirms it gets the value back out properly.

File: tests/hazmat/backends/test_openssl.py (hunk 8 above)
Comment (reaperhulk) [inline tests/hazmat/backends/test_openssl.py]: Also, this should probably be secp256r1 as it's a more widely supported curve across the openssl versions we support.

**Comment by alex** [comment]
As a sort of general point, I think the test cases should be pruned down, to only set the attribtues that are relevant to the tests.

That said, I think this is merge-ready. @reaperhulk I'll let you double check.

**Comment by alex** [comment]
Ugh, I lied: https://coveralls.io/builds/2899126 there's a coverage issue

**Comment by alex** [comment]
Missing test cases:

- `CSRBuilder().subject_name(x509.Name()).subject_name(x509.Name())`
- `backend.create_x509_csr` where it raises `NotImplementedError` on unknown extension.
  - This is a place where the logic is handled by the frontend already; perhaps we can move all support OID checks to the backend?

File: src/cryptography/hazmat/backends/interfaces.py [hunk 9]
```diff
@@ -274,6 +274,12 @@ def load_pem_x509_csr(self, data):
         Load an X.509 CSR from PEM encoded data.
         """
 
+    @abc.abstractmethod
+    def create_x509_csr(self, builder, private_key, algorithm):
```
Comment (reaperhulk) [inline src/cryptography/hazmat/backends/interfaces.py]: This needs a quick doc in the backend interfaces.

**Comment by reaperhulk** [comment]
Other than my one comment this LGTM. Awesome work @AndreLouisCaron and @sigmavirus24!
